File: src/common/web/api/project/ProjectExportersCommands.ts
import { Type } from "class-transformer";

import { Command } from "../../core/messaging/Command";
import { CommandReply } from "../../core/messaging/CommandReply";
import { CommandComposer } from "../../core/messaging/composers/CommandComposer";
import { CommandReplyComposer } from "../../core/messaging/composers/CommandReplyComposer";
import { MessageBuilder } from "../../core/messaging/composers/MessageBuilder";
import { Message } from "../../core/messaging/Message";
import { type ProjectFeatureID } from "../../data/entities/project/features/ProjectFeature";
import { type ProjectID } from "../../data/entities/project/Project";
import { ProjectExporterDescriptor, type ProjectExporterID } from "../../data/exporters/ProjectExporterDescriptor";

/**
 * Command to fetch all project exporters. Requires a ``ListProjectExportersReply`` reply.
 */
@Message.define("command/project/export/list")
export class ListProjectExportersCommand extends Command {
    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, chain: Message | null = null): CommandComposer<ListProjectExportersCommand> {
        return messageBuilder.buildCommand(ListProjectExportersCommand, {}, chain);
    }
}

/**
 * Reply to ``Reply to ``ListProjectsCommand``.``.
 *
 * @param exporters - List of all project exporters.
 */
@Message.define("command/project/export/list/reply")
export class ListProjectExportersReply extends CommandReply {
    // @ts-ignore
    @Type(() => ProjectExporterDescriptor)
    public readonly exporters: ProjectExporterDescriptor[] = [];

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: ListProjectExportersCommand,
        exporters: ProjectExporterDescriptor[],
        success: boolean = true,
        message: string = ""
    ): CommandReplyComposer<ListProjectExportersReply> {
        return messageBuilder.buildCommandReply(ListProjectExportersReply, cmd, success, message, { exporters: exporters });
    }
}

/**
 * Command to export a project. Requires an ``ExportProjectReply`` reply.
 */
@Message.define("command/project/export")
export class ExportProjectCommand extends Command {
    public readonly project_id: ProjectID = 0;

    public readonly exporter: ProjectExporterID = "";
    public readonly scope: ProjectFeatureID = "";

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        projectID: ProjectID,
        exporter: ProjectExporterID,
        scope: ProjectFeatureID,
        chain: Message | null = null
    ): CommandComposer<ExportProjectCommand> {
        return messageBuilder.buildCommand(ExportProjectCommand, { project_id: projectID, exporter: exporter, scope: scope }, chain);
    }
}

/**
 * Reply to ``ExportProjectCommand``.
 *
 * @param mimetype - The MIME type of the export result.
 */
@Message.define("command/project/export/reply")
export class ExportProjectReply extends CommandReply {
    public readonly mimetype: string = "";

    /**
     * The data of the export result.
     */
    public get data(): ArrayBuffer | undefined {
        if (this.payload.contains("data")) {
            return this.payload.get("data") as ArrayBuffer;
        }
        return undefined;
    }

    /**
     * Sets the data of the export result.
     *
     * @param data - The export result data.
     */
    public set data(data: ArrayBuffer) {
        this.payload.set("data", data);
    }

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: ExportProjectCommand,
        mimetype: string,
        data: ArrayBuffer,
        success: boolean = true,
        message: string = ""
    ): CommandReplyComposer<ExportProjectReply> {
        const composer = messageBuilder.buildCommandReply(ExportProjectReply, cmd, success, message, { mimetype: mimetype });
        composer.addPayload("data", data);
        return composer;
    }
}

File: src/common/py/api/project/project_exporters_commands.py
import dataclasses
import typing

from ...core.messaging import Command, CommandReply, Message
from ...core.messaging.composers import (
    CommandComposer,
    CommandReplyComposer,
    MessageBuilder,
)
from ...data.entities.project import ProjectID
from ...data.entities.project.features import ProjectFeatureID
from ...data.exporters import (
    ProjectExporterDescriptor,
    ProjectExporterID,
    ProjectExporterResult,
)


@Message.define("command/project/export/list")
class ListProjectExportersCommand(Command):
    """
    Command to fetch all project exporters.

    Notes:
        Requires a ``ListProjectExportersReply`` reply.
    """

    @staticmethod
    def build(
        message_builder: MessageBuilder, *, chain: Message | None = None
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(ListProjectExportersCommand, chain)


@Message.define("command/project/export/list/reply")
class ListProjectExportersReply(CommandReply):
    """
    Reply to ``ListProjectExportersCommand``.

    Args:
        exporters: List of all project exporters.
    """

    exporters: typing.List[ProjectExporterDescriptor] = dataclasses.field(
        default_factory=list
    )

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: ListProjectExportersCommand,
        *,
        exporters: typing.List[ProjectExporterDescriptor],
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            ListProjectExportersReply, cmd, success, message, exporters=exporters
        )


@Message.define("command/project/export")
class ExportProjectCommand(Command):
    """
    Command to export a project.

    Notes:
        Requires an ``ExportProjectReply`` reply.
    """

    project_id: ProjectID

    exporter: ProjectExporterID
    scope: ProjectFeatureID

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        project_id: ProjectID,
        exporter: ProjectExporterID,
        scope: ProjectFeatureID,
        chain: Message | None = None,
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(
            ExportProjectCommand,
            chain,
            project_id=project_id,
            exporter=exporter,
            scope=scope,
        )


@Message.define("command/project/export/reply")
class ExportProjectReply(CommandReply):
    """
    Reply to ``ExportProjectCommand``.

    Args:
        mimetype: The MIME type of the export result.
    """

    mimetype: str

    @property
    def data(self) -> bytes | None:
        """
        The data of the export result.

        Returns:
            The binary data of the export result.
        """
        return self.payload.get("data") if "data" in self.payload else None

    @data.setter
    def data(self, data: bytes) -> None:
        """
        Sets the data of the export result.

        Args:
            data: The export result data.
        """
        self.payload.set("data", data)

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: ExportProjectCommand,
        *,
        mimetype: str,
        data: bytes,
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        composer = message_builder.build_command_reply(
            ExportProjectReply, cmd, success, message, mimetype=mimetype
        )
        composer.add_payload("data", data)
        return composer
